### ISSUE TITLE
Special Test Request: Handling of connection problems

### ISSUE THREAD

**Issue by flogat** [shell-fire/shellfire-vpn#41]
Please test the following scenarios with an updated build for proper handling of the app in these circumstances:

Activities:
- start of app
- logging in
- requesting password
- registering
- changing server
- changing protocol
- connecting
- disconnecting

each in the following scenarios:
1 connection available
2 connection down (e.g. disable network adapter)
3 connection timeout to webservice/api (e.g. if endpoint=www.shellfire.de -> put in hosts file: "1.3.4.5	www.shellfire.de"
4 connection timeout to vpn server (e.g. if vpn-server is blubb.shellfire-vpn.com -> put in hosts file: "1.2.3.4 blubb.shellfire-vpn.com"

(4 only for connecting/disconnect)

Note: The changes to host file and disabling of network adapter need to be made "on the fly", e.g. after start of app, logging in, etc. to simulate unavailable network, blocked dns, etc.

**Comment by mbehrsf** [comment]
Here you go:

- Start of app
1. OK
2. OK - App displays "no internet connection" and closes
3. OK (looks for endpoints, then starts)

- logging in
1. OK
2. NOT OK - Displays "unknown login error", but the Login status indicator stays there and keeps moving
3. NOT OK (Login gives unknown error)

- Requesting password
1. OK
2. OK (browser does not load, of course)
3. OK (browser does not load because www.shellfire.de is blocked)

- Registering
1. NOT OK (see https://github.com/shell-fire/shellfire-vpn/issues/37)
2. NOT OK (see  https://github.com/shell-fire/shellfire-vpn/issues/37)
3. NOT OK (see  https://github.com/shell-fire/shellfire-vpn/issues/37)

- Changing server / protocol (I am quite sure it's technically the same thing?)
1. OK
2. NOT OK (no error message or anything)
3. OK (switches to different endpoint, nice!)

- Connecting
1. OK
2. NOT OK (Says Connecting forever, I waited quite long, looks like there is no timeout?)
3. NOT OK (Cannot connect, cannot send log)
4. NOT OK (Connection fails, recommends reboot, should probably recommend checking the internet connection first?)

- Disconnecting
1. OK
2. OK
3. OK (Connection goes through the VPN)
4. OK (ReachableWithTimeout counts to 5, tries to reach google, counts to 2 again, then disconnects and attempts reconnect)


-

**Comment by flogat** [comment]
@mbehrsf 
- Added a generic "Unknown error" for the cases when a setServerTo is tried to be performed, but backend is not available (e.g. no connection at all or backend dns-blocked)
- As it would be some effort to introduce custom error messages in these cases (e.g. "No Connection" or "Endpoint not available" or even try to automatically find another endpoint), I think this generic error is quite okay - do you agree?
- I am wondering how you blocked the vpn server - we're only using IPs, so I could not reproduce the 4. case myself - how did you do it?

Other than that I am quite happy with the most recent version - it's rebuilt and uploaded, please check.

**Comment by mbehrsf** [comment]
> * I am wondering how you blocked the vpn server - we're only using IPs, so I could not reproduce the 4. case myself - how did you do it?
> 
I used a Shellfire Box and added custom iptables rules in LuCi:
iptables -I FORWARD --destination [IP ADRESS] -j REJECT 

Testing the new build atm

**Comment by mbehrsf** [comment]
I agree about the generic error message being enough, but it won't show up in most cases. Also, in case 3, the app should try switching to a different endpoint first before giving up and throwing an error.

Re-tested NOT OK cases, 

- logging in

2. OK (Unknown error but no eternal loading indicator)
3. NOT OK (Unknown error but no eternal loading indicator, I guess it should search for different endpoints first before cancelling?)

- Registering
NOT OK (see #37)
NOT OK (see #37)
NOT OK (see #37)

- Changing server / protocol 
2. NOT OK (no error message or anything, just eternal "Connecting")

- Connecting
2. NOT OK (Sometimes it says Connecting forever, I waited quite long, looks like there is no timeout? Other times I get Unknown error)
3. NOT OK (Unknown error, not trying to switch endpoints)
4. NOT OK (Eternal connect this time, no error like with the previous build)

**Comment by flogat** [comment]
Okay, so after thinking about it some more:
- Once a valid api endpoint is found after startup and loading of login-form, I cannot really think of a practical usecase, where the api would not be available anymore, but another endpoint would work. Sorry if I caused extra work here, but I guess these tests (number 3) should be removed from the scope.

Remaining cases:
- Registering / all - discussed in separate #37 
- Changing server / protocol / connecting -> Loss of connection (e.g. disabled network adapter) -> #50 
- Connecting to vpn server where name does not work - #51
